### PR TITLE
Add-mini-block-partial-execution-support

### PR DESCRIPTION
**Pull request by SebastianMarian**
* Added mini block partial execution support in mini block header Reserved field

### CODE REVIEW

File: data/block/miniBlockHeader_test.go [hunk 1]
```diff
@@ -52,14 +52,34 @@ func TestMiniBlockHeader_IsFinalFromConstructionState(t *testing.T) {
...
+	mbhReserved := block.MiniBlockHeaderReserved{IndexOfLastTxProcessed: 78}
+	reserved, err := mbhReserved.Marshal()
+	mbh.Reserved = reserved
+	require.Nil(t, err)
+	index = mbh.GetIndexOfLastTxProcessed()
+	require.Equal(t, int32(78), index)
+
+	mbhReserved.IndexOfLastTxProcessed = 21
+	mbh.Reserved, _ = mbhReserved.Marshal()
+	require.Equal(t, int32(21), mbh.GetIndexOfLastTxProcessed())
```
Comment (sstanculeanu) [inline data/block/miniBlockHeader_test.go:72]:
```suggestion
	require.Equal(t, mbhReserved.IndexOfLastTxProcessed, mbh.GetIndexOfLastTxProcessed())
```

File: data/block/miniBlockHeader_test.go [hunk 2]
```diff
@@ -52,14 +52,34 @@ func TestMiniBlockHeader_IsFinalFromConstructionState(t *testing.T) {
 	require.True(t, mbh.IsFinal())
 }
 
+func TestMiniBlockHeader_GetIndexOfLastTxProcessed(t *testing.T) {
+	mbh := &block.MiniBlockHeader{
+		TxCount: 70,
+	}
+
+	index := mbh.GetIndexOfLastTxProcessed()
+	require.Equal(t, int32(69), index)
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply data/block/miniBlockHeader_test.go:61]: Done

File: data/block/miniBlockHeader_test.go (hunk 1 above)
Comment (SebastianMarian) [reply data/block/miniBlockHeader_test.go:72]: Done

File: data/block/miniBlockHeader_test.go [hunk 3]
```diff
@@ -52,14 +52,34 @@ func TestMiniBlockHeader_IsFinalFromConstructionState(t *testing.T) {
...
+
+	mbhReserved := block.MiniBlockHeaderReserved{IndexOfLastTxProcessed: 78}
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply data/block/miniBlockHeader_test.go:63]: Refactored

File: data/block/miniBlockHeader_test.go [hunk 4]
```diff
@@ -68,48 +88,129 @@ func TestMiniBlockHeader_SetConstructionState(t *testing.T) {
...
-	require.Equal(t, int32(block.Final), mbh.GetConstructionState())
+	_ = mbh.SetConstructionState(int32(block.PartialExecuted))
+	require.Equal(t, int32(block.PartialExecuted), mbh.GetConstructionState())
 }
 
-func TestMiniBlockHeader_SetProcessingTypeDoesNotChangeConstructionState(t *testing.T) {
+func TestMiniBlockHeader_SetIndexOfLastTxProcessed(t *testing.T) {
+	mbh := &block.MiniBlockHeader{}
+
+	_ = mbh.SetIndexOfLastTxProcessed(3)
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply data/block/miniBlockHeader_test.go:98]: Done

File: data/block/miniBlockHeader_test.go [hunk 5]
```diff
@@ -68,48 +88,129 @@ func TestMiniBlockHeader_SetConstructionState(t *testing.T) {
...
+	require.Equal(t, int32(3), mbh.GetIndexOfLastTxProcessed())
+
+	_ = mbh.SetIndexOfLastTxProcessed(5)
+	require.Equal(t, int32(5), mbh.GetIndexOfLastTxProcessed())
+}
+
+func TestMiniBlockHeader_SetIndexOfFirstTxProcessed(t *testing.T) {
+	mbh := &block.MiniBlockHeader{}
+
+	_ = mbh.SetIndexOfFirstTxProcessed(3)
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply data/block/miniBlockHeader_test.go:108]: Done

**Review by sstanculeanu — DISMISSED** [review]

File: data/block/blockV2.proto [hunk 6]
```diff
@@ -28,4 +28,6 @@ message MiniBlockReserved {
 message MiniBlockHeaderReserved {
   ProcessingType ExecutionType = 1;
   MiniBlockState State = 2;
+  int32 IndexOfFirstTxProcessed = 3;
```
Comment (bogdan-rosianu) [inline data/block/blockV2.proto:31]: why int32 instead of uint32?

File: data/block/miniBlockHeader.go [hunk 7]
```diff
@@ -164,6 +164,56 @@ func (m *MiniBlockHeader) setMiniBlockHeaderReserved(mbhr *MiniBlockHeaderReserv
...
+	}
+
+	return miniBlockHeaderReserved.IndexOfFirstTxProcessed
+}
+
+// GetIndexOfLastTxProcessed returns index of the last transaction processed in the miniBlock
+func (m *MiniBlockHeader) GetIndexOfLastTxProcessed() int32 {
+	miniBlockHeaderReserved, err := m.getMiniBlockHeaderReserved()
+	if err != nil || miniBlockHeaderReserved == nil {
+		return int32(m.TxCount) - 1
```
Comment (bogdan-rosianu) [inline data/block/miniBlockHeader.go:181]: if 0 tx count you would get -1. expected? do you treat this when used?

File: data/block/blockV2.proto (hunk 6 above)
Comment (bogdan-rosianu) [reply data/block/blockV2.proto:31]: Do you check somewhere in the code for `-1` values? if yes, then it's fine

File: data/block/miniBlockHeader_test.go [hunk 8]
```diff
@@ -68,48 +89,133 @@ func TestMiniBlockHeader_SetConstructionState(t *testing.T) {
...
+}
+
+func TestMiniBlockHeader_SetIndexOfTxProcessedDoesNotChangeConstructionStateOrProcessingType(t *testing.T) {
+	mbh := &block.MiniBlockHeader{}
+	processingType := int32(block.Scheduled)
+	_ = mbh.SetProcessingType(processingType)
+	state := int32(block.Proposed)
+	_ = mbh.SetConstructionState(state)
+
+	_ = mbh.SetIndexOfFirstTxProcessed(20)
```
Comment (bogdan-rosianu) [inline data/block/miniBlockHeader_test.go:197]: please also add tests for the bounders (index 0, index -1, index max int32, and so on)

File: data/block/blockV2.proto (hunk 6 above)
Comment (SebastianMarian) [reply data/block/blockV2.proto:31]: Yes a cross mini block not executed yet (pending) has index of last tx processed -1

File: data/block/miniBlockHeader.go (hunk 7 above)
Comment (SebastianMarian) [reply data/block/miniBlockHeader.go:181]: There could not exists empty mini blocks added in block body. At least 1 tx should be in each mini block added by proposer in block body and signed by consensus group as well.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

File: data/block/miniBlockHeader_test.go (hunk 8 above)
Comment (SebastianMarian) [reply data/block/miniBlockHeader_test.go:329]: Done

**Review by iulianpascalau — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**